src/errors/mod.rs

```diff
@@ -0,0 +1,41 @@
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+pub enum PSError {
```

Comment:
I think its a bit confusing having errors from the ode solvers, linear solver and nonlinear solvers all in the one enum. Perhaps break it up so that you have `OdeSolverError` for the `OdeSolverMethod` trait, `LinearSolverError` for the `LinearSolver` trait, `NonLinearSolverError` for the `NonLinearSolver` trait etc.

Reply:
I agree - and that's what I started with. However, some operations have errors that are the same, e.g. some errors in `LinearSolverError` can also be in `NonLinearSolverError`. But I guess its a low price to pay for additional granularity.

Reply:
yea, I'm happy with some repeated enum cases between `LinearSolverError` and `NonLinearSolverError`